### PR TITLE
Enhance/new model take2

### PR DESCRIPTION
**Pull request by mat10tng**
Ok, lots of changes due to merge in from master . So, I will list the file that are important to look at:

We defined the schema here
- src/app/model/schema.ts
My base model:
- src/app/model/BaseModel.ts
we have a few helper methods there to load, and sanitize incoming data

Model implemented on
- src/app/model/User.ts
- src/app/model/Mission.ts

Specifically, the User model now have a few methods, such as
volunteerMission, startMission and deliverMission

These methods are then use in these pages
src/app/page/Missions.jsx -> volunteerMission() ->
src/app/page/MissionAssigned.jsx -> startMission()->
src/app/page/MissionsStarted.jsx -> deliverMission()->

I have been a bit lazy here and only apply loads on the 
src/app/page/Missions.jsx 
in const mapStateToProps line 67
```
const mapStateToProps = (state, ownProps) => {
  let missions = state.firestore.ordered.missionsUnassigned;
  missions = Mission.loads(missions);
  return {
    missions: missions,
    auth: state.firebase.auth,
  };
};
```
Eventually, when update data into the firebase, we would want to use sanitize function as well. 
But since this is a big changes, I will leave it at that for now

Finnaly:
added a few new pages

https://tinyurl.com/yarmzb32
https://tinyurl.com/y8rluagd
https://tinyurl.com/yb6toj7t
https://tinyurl.com/y85jbgke

### CODE REVIEW

**Comment by mat10tng** [comment]
https://tinyurl.com/yarmzb32
https://tinyurl.com/y8rluagd
https://tinyurl.com/yb6toj7t
https://tinyurl.com/y85jbgke

image
image

image

image

File: src/app/model/BaseModel.ts [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    return this.fs.collection(collectionName);
+  }
+
+  /**
+   * @param data: the data to load in and create a model object
+   * all fields outside of the schema will not be loaded in
+   * @returns: the object with all fields
+   */
+  load(data: unknown) {
+    return data ? merge(data, this.defaultData) : data;
```
Comment (qdozaq) [inline src/app/model/BaseModel.ts:41]: Do we not want the `defaultData` returned here if there is no data?

File: src/app/model/BaseModel.ts [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    });
+  }
+
+  /**
+   * @param data: the data to load in and create a model object
+   * all fields outside of the schema will not be loaded in
+   * @returns: the object with all fields sanitized
+   */
+  sanitize(data: unknown) {
+    return data ? sanitize(data, this.defaultData) : data;
```
Comment (qdozaq) [inline src/app/model/BaseModel.ts:62]: We definitely only want to return `return sanitize(...)` so we know the data is properly sanitized.
This works for `undefined`
but if the value is `false` `null` or `""` then we pass it through without checking if that was the correct type for that object.

File: src/app/model/Mission.ts [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+  MissionFundedStatus,
+  MissionPayableStatus,
+  MissionDetails,
+} from "./schema";
+import BaseModel from "./BaseModel";
+
+const defaultLocation: Location = {
+  address: "",
+  lat: 0,
+  long: 0,
```
Comment (qdozaq) [inline src/app/model/Mission.ts:17]: does this need to be `lng`? Or did you fix that?

File: src/app/model/User.ts [hunk 4]
```diff
@@ -0,0 +1,107 @@
+import { UserInterface, VolunteerStatus, Location, MissionStatus } from "./schema";
+import BaseModel from "./BaseModel";
+
+const defaultLocation: Location = {
```
Comment (qdozaq) [inline src/app/model/User.ts:4]: Wonder if we should just define the defaults along side the types. Then we could share them like we do the types

File: src/app/model/User.ts [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+  VolunteerStatus = VolunteerStatus;
+
+  /**
+   * User volunteer for a mission
+   * @param {string} userId : user
+   * @param {string} missionId : mission that user want to volunteer for
+   */
+  async volunteerMission(userId: string, missionId: string) {
+    let collection = this.getCollection("missions");
+    let doc = await collection.doc(missionId).get();
```
Comment (qdozaq) [inline src/app/model/User.ts:37]: Wrap in try/catch?

File: src/app/model/User.ts [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+  }
+
+  /**
+   * User start a mission
+   * @param {string} userId - user
+   * @param {string} missionId - mission that user want to start
+   */
+  async startMission(userId: string, missionId: string) {
+    let collection = this.getCollection("missions");
+    let doc = await collection.doc(missionId).get();
```
Comment (qdozaq) [inline src/app/model/User.ts:64]: same here

File: src/app/model/User.ts [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+
+  /**
+   * User deliver a mission
+   * //TOD add image
+   * @param {string} userId - user
+   * @param {string} missionId  - mission id that user delivered
+   */
+  async deliverMission(userId: string, missionId: string) {
+    let collection = this.getCollection("missions");
+    let doc = await collection.doc(missionId).get();
```
Comment (qdozaq) [inline src/app/model/User.ts:89]: and here

File: src/app/model/schema.ts [hunk 8]
```diff
@@ -104,16 +103,16 @@ export enum MissionType {
...
-interface FoodBoxDetails extends MissionDetails {
+export interface FoodBoxDetails extends MissionDetails {
   boxes?: Array<FoodBox>; // if multple boxes of same type grouping can occur in the UI
 }
 
 // created by the organiser.. but for MVP.0 we *hard code* a list of one type of box
-@Collection("foodboxes")
-export class FoodBox {
-  id!: string;
+// @Collection("foodboxes")
```
Comment (qdozaq) [inline src/app/model/schema.ts:113]: We're not using the `@Collection` tags anymore right?

File: src/index.tsx [hunk 9]
```diff
@@ -22,7 +22,7 @@ const rrfConfig = {
 
 firebase.initializeApp(firebaseConfig);
 const firestore = firebase.firestore();
-fireorm.initialize(firestore);
+//fireorm.initialize(firestore);
```
Comment (qdozaq) [inline src/index.tsx:25]: delete?

File: src/app/model/Mission.ts (hunk 3 above)
Comment (mat10tng) [reply src/app/model/Mission.ts:17]: yeah, it need to lng

File: src/app/model/BaseModel.ts (hunk 1 above)
Comment (mat10tng) [reply src/app/model/BaseModel.ts:41]: no, I will remove this

File: src/app/model/BaseModel.ts (hunk 2 above)
Comment (mat10tng) [reply src/app/model/BaseModel.ts:62]: fixed

File: src/app/model/schema.ts (hunk 8 above)
Comment (mat10tng) [reply src/app/model/schema.ts:113]: removed

File: package.json [hunk 10]
```diff
@@ -13,6 +13,7 @@
     "@reduxjs/toolkit": "^1.3.4",
     "algolia-places-react": "^1.5.1",
     "class-validator": "^0.11.1",
+    "dataclass": "^1.2.0",
```
[the remark this answers is not in the document]
Comment (mat10tng) [reply package.json:16]: yes, on it

File: src/app/model/User.ts (hunk 4 above)
Comment (mat10tng) [reply src/app/model/User.ts:4]: we can wait with that for now, I do not expect defaultData to be touch from outside

File: src/app/model/User.ts (hunk 5 above)
Comment (mat10tng) [reply src/app/model/User.ts]: done

File: src/app/model/User.ts (hunk 7 above)
Comment (mat10tng) [reply src/app/model/User.ts]: done

File: src/index.tsx (hunk 9 above)
Comment (mat10tng) [reply src/index.tsx]: done

**Pull request merged**